### PR TITLE
[libra config] emit NewEpochEvent in LibraConfig

### PR DESCRIPTION
**Pull request by zekun000**
## Motivation
Another step towards moving specialized ValidatorChangeEvent to the
general LibraConfig module.

They'll co-exist for this commit and merged in next commit.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

Existing tests.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

File: execution/executor/src/lib.rs [hunk 1]
```diff
@@ -678,9 +681,22 @@ where
             next_validator_set = vm_output
                 .events()
                 .iter()
-                .find(|event| *event.key() == validator_set_change_event_key)
-                .map(|event| ValidatorSet::from_bytes(event.event_data()))
-                .transpose()?
+                .find(|event| {
+                    *event.key() == validator_set_change_event_key
```
Comment (runtian-zhou) [inline execution/executor/src/lib.rs:685]: cc @christinacelee

File: language/stdlib/modules/libra_configs.move [hunk 2]
```diff
@@ -1,32 +1,104 @@
...
         // Only callable by the Association address for now.
         Transaction::assert(Transaction::sender() == 0xA550C18, 1);
 
         move_to_sender(T{ payload });
-        LibraSystem::reconfigure();
+        reconfigure();
+    }
+
+
+    fun reconfigure() acquires Configuration {
```
Comment (runtian-zhou) [inline language/stdlib/modules/libra_configs.move:73]: `reconfigure` should probably be public. One example is that `WriteSet` transaction needs to call this function to trigger a new epoch.

File: language/stdlib/modules/libra_system.move [hunk 3]
```diff
@@ -16,8 +16,7 @@ module LibraSystem {
     }
 
     struct ValidatorSetChangeEvent {
-        scheme: u8,
-        new_validator_set: vector<ValidatorInfo>,
+        last_reconfiguration_time: u64,
```
Comment (runtian-zhou) [inline language/stdlib/modules/libra_system.move:19]: Are we going to get rid of `last_reconfiguration_time` in `libra_system` in the PRs to come?

Comment (zekun000) [reply language/stdlib/modules/libra_system.move:19]: yup

File: execution/executor/src/lib.rs [hunk 4]
```diff
@@ -678,9 +681,22 @@ where
...
+                        || *event.key() == new_epoch_event_key
```
Comment (davidiw) [inline execution/executor/src/lib.rs:686]: isn't new_epoch_event_key sufficient? isn't one event a superset of the other?

File: language/stdlib/modules/libra_configs.move [hunk 5]
```diff
@@ -1,32 +1,101 @@
...
+
+        move_to_sender<Configuration>(Configuration {
+            epoch: 0,
+            last_reconfiguration_time: 0,
+            events: LibraAccount::new_event_handle<NewEpochEvent>(),
+        });
+    }
+
+    // We can't return reference here, this returns a copy of the config value.
+    public fun get<Config: copyable>(): Config acquires T {
```
Comment (davidiw) [inline language/stdlib/modules/libra_configs.move:35]: I don't think we need permission checks on retrieving copies

File: language/stdlib/modules/libra_configs.move [hunk 6]
```diff
@@ -1,32 +1,101 @@
...
         let sender = Transaction::sender();
 
         // Only callable by the Association address for now.
         Transaction::assert(sender == 0xA550C18, 1);
 
+        Transaction::assert(::exists<T<Config>>(sender), 24);
+        let t_ref = borrow_global_mut<T<Config>>(sender);
```
Comment (davidiw) [inline language/stdlib/modules/libra_configs.move:42]: This should be a lint error, but you're borrowing a mut and only using it as read-only

File: language/stdlib/modules/libra_configs.move [hunk 7]
```diff
@@ -1,32 +1,101 @@
...
+
+        *&t_ref.payload
+    }
+
+    // Set a config item to a new value and trigger a reconfiguration.
+    // This can only be invoked by the Association address
+    public fun set<Config: copyable>(payload: Config) acquires T, Configuration {
+        let sender = Transaction::sender();
+
```
Comment (davidiw) [inline language/stdlib/modules/libra_configs.move:51]: here you need the permission checks

Comment (zekun000) [reply language/stdlib/modules/libra_configs.move:51]: permission check is moved to reconfigure function

File: execution/executor/src/lib.rs (hunk 4 above)
Comment (zekun000) [reply execution/executor/src/lib.rs:686]: no, next commit will remove validator_change_event. currently it's not part of the libra config yet

File: language/stdlib/modules/libra_configs.move (hunk 7 above)
Comment (davidiw) [reply language/stdlib/modules/libra_configs.move:48]: Oh, I see, you are only allowed to call set if you have this resource... asserting that the resource exists (49) is interesting... what happens if we delete it, have no such resource, and call config.payload

File: language/stdlib/modules/libra_configs.move (hunk 5 above)
Comment (davidiw) [reply language/stdlib/modules/libra_configs.move:35]: It seems interesting that get and set allow for anyone to host this, but reconfigure enforces only the association can. Should we make get read only from the association account? For symmetry or leave a comment?

File: language/stdlib/modules/libra_configs.move (hunk 7 above)
Comment (zekun000) [reply language/stdlib/modules/libra_configs.move:48]: it'll abort, we don't support delete right now, but I imagine if we delete it, it has to go through `publish_new_config`

**Comment by zekun000** [comment]
review can move to https://github.com/libra/libra/pull/3308 directly

**Pull request closed without merging**